server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java

```diff
@@ -0,0 +1,215 @@
+
+    private static final ParseField FIELD_FIELD = new ParseField("field");
+    private static final ParseField ORIGIN_FIELD = new ParseField("origin");
+    private static final ParseField PIVOT_FIELD = new ParseField("pivot");
+
+    private final String field;
+    private final Origin origin;
+    private final String pivot;
+
+    private static final ConstructingObjectParser<DistanceFeatureQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        "distance_feature", false,
+        args -> new DistanceFeatureQueryBuilder((String) args[0], (Origin) args[1], (String) args[2])
+    );
+
+    static {
+        PARSER.declareString(constructorArg(), FIELD_FIELD);
+        // origin: number or string for date and date_nanos fields; string, array, object for geo fields
+        PARSER.declareField(constructorArg(), DistanceFeatureQueryBuilder.Origin::originFromXContent,
+            ORIGIN_FIELD, ObjectParser.ValueType.OBJECT_ARRAY_STRING_OR_NUMBER);
+        PARSER.declareString(constructorArg(), PIVOT_FIELD);
+        declareStandardFields(PARSER);
+    }
+
+    public DistanceFeatureQueryBuilder(String field, Origin origin, String pivot) {
+        this.field = Objects.requireNonNull(field);
+        this.origin = Objects.requireNonNull(origin);
+        this.pivot = Objects.requireNonNull(pivot);
+    }
+
+    public static DistanceFeatureQueryBuilder fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(FIELD_FIELD.getPreferredName(), field);
+        builder.field(ORIGIN_FIELD.getPreferredName(), origin.origin);
+        builder.field(PIVOT_FIELD.getPreferredName(), pivot);
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+
+    public DistanceFeatureQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        field = in.readString();
+        origin = new Origin(in);
+        pivot = in.readString();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(field);
+        origin.writeTo(out);
+        out.writeString(pivot);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        MappedFieldType fieldType = context.fieldMapper(field);
+        if (fieldType == null) {
+            throw new IllegalArgumentException("Can't run [" + NAME + "] query on unmapped fields!");
+        }
+        Object originObj = origin.origin();
+        if (fieldType instanceof DateFieldType) {
+            long originLong = (originObj instanceof Long) ? (Long) originObj :
+                ((DateFieldType) fieldType).parseToLong(originObj, true, null, null, context);
```

Comment:
I don't like taking longs as-is, we should always go through `parseToLong` imo. I'd like to avoid that users need to be aware of the internal resolution of the field.

Reply:
@jpountz  `parseToLong` does NOT work for date_nanos field when parsing a huge long value (an expected value for `date_nanos`). 
`DateFieldType::parseToLong`  uses `JavaDateMathParser` (instead of desired `JodaDateMathParser`) even if a  date type is `date_nanos`. Or should I pass a `forcedDateParser` to `parseToLong` depending on the fieldType's resolution?

Reply:
@jpountz  Please disregard my previous comment here. I assumed that `date_nanos` except a long value as nanoseconds-since-the-epoch, but actually the long value should be  milliseconds-since-the-epoch.   
I have made the changes as you suggested.